sw/device/tests/dif/dif_gpio_smoketest.c

```diff
@@ -0,0 +1,66 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_gpio.h"
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"  // Generated.
+
+static dif_gpio_t gpio;
+const test_config_t kTestConfig;
+
+// Pins 20-23 are grounded and pins 16-19 are reserved for JTAG.
+const uint32_t kGpioMask = 0xFF00FFFF;
+
+/**
+ * Writes the given value to GPIO pins and compares it against the value read.
+ *
+ * Masks the bits that correspond to the pins that we cannot test.
+ *
+ * See also: `kGpioMask`.
+ *
+ * @param write_val Value to write.
+ */
+static void test_gpio_write(uint32_t write_val) {
+  CHECK(dif_gpio_write_all(&gpio, write_val) == kDifGpioOk);
+
+  uint32_t read_val = 0;
+  CHECK(dif_gpio_read_all(&gpio, &read_val) == kDifGpioOk);
+
+  uint32_t expected = write_val & kGpioMask;
+  uint32_t actual = read_val & kGpioMask;
+  CHECK(expected == actual, "%X != %X", expected, actual);
+}
+
+/**
+ * Smoke test for the GPIO peripheral.
+ *
+ * Performs a loopback test by writing various values and reading them back.
+ */
+bool test_main(void) {
+  CHECK(dif_gpio_init(
+            (dif_gpio_params_t){
+                .base_addr = mmio_region_from_addr(TOP_EARLGREY_GPIO_BASE_ADDR),
+            },
+            &gpio) == kDifGpioOk);
+  CHECK(dif_gpio_output_set_enabled_all(&gpio, kGpioMask) == kDifGpioOk);
+
+  const uint32_t kVals[] = {0xAAAAAAAA, 0x55555555, 0xA5A5A5A5, 0xFFFFFFFF, 0};
+  for (uint8_t i = 0; i < ARRAYSIZE(kVals); ++i) {
+    test_gpio_write(kVals[i]);
+  }
+
+  // Walking 1s and 0s.
+  for (uint32_t i = 1; i > 0; i <<= 1) {
```

Comment:
Haha, this is pretty clever, in a good way. I would maybe update the comment to be more specific about the iteration sequence? E.g. "iterate over every integer with one bit set and every integer with all but one bit set".

Reply:
Done.